pkg/providers/tinkerbell/reconciler/reconciler.go

```diff
@@ -117,17 +139,97 @@ func (r *Reconciler) ValidateClusterSpec(ctx context.Context, log logr.Logger, t
 	return controller.Result{}, nil
 }
 
-// ReconcileControlPlane applies the control plane CAPI objects to the cluster.
-func (r *Reconciler) ReconcileControlPlane(ctx context.Context, log logr.Logger, tinkerbellScope *Scope) (controller.Result, error) {
+// GenerateSpec generates Tinkerbell control plane and workers spec.
+func (r *Reconciler) GenerateSpec(ctx context.Context, log logr.Logger, tinkerbellScope *Scope) (controller.Result, error) {
 	spec := tinkerbellScope.ClusterSpec
-	log = log.WithValues("phase", "reconcileControlPlane")
-	log.Info("Applying control plane CAPI objects")
+	log = log.WithValues("phase", "GenerateSpec")
+
 	cp, err := tinkerbell.ControlPlaneSpec(ctx, log, clientutil.NewKubeClient(r.client), spec)
+	if err != nil {
+		return controller.Result{}, errors.Wrap(err, "generating control plane spec")
+	}
+	tinkerbellScope.ControlPlane = cp
+
+	w, err := tinkerbell.WorkersSpec(ctx, log, clientutil.NewKubeClient(r.client), spec)
+	if err != nil {
+		return controller.Result{}, errors.Wrap(err, "generating workers spec")
+	}
+	tinkerbellScope.Workers = w
+
+	return controller.Result{}, nil
+}
+
+// DetectOperation detects change type.
+func (r *Reconciler) DetectOperation(ctx context.Context, log logr.Logger, tinkerbellScope *Scope) (Operation, error) {
+	log.Info("Detecting operation type")
+
+	currentKCP, err := controller.GetKubeadmControlPlane(ctx, r.client, tinkerbellScope.ClusterSpec.Cluster)
+	if err != nil {
+		return "", err
+	}
+	if currentKCP == nil {
+		return NewClusterOperation, nil
+	}
+
+	wantVersionChange := currentKCP.Spec.Version != tinkerbellScope.ControlPlane.KubeadmControlPlane.Spec.Version
+	cpWantScaleChange := *currentKCP.Spec.Replicas != *tinkerbellScope.ControlPlane.KubeadmControlPlane.Spec.Replicas
+	workerWantScaleChange := r.workerReplicasDiff(ctx, tinkerbellScope)
+	wantScaleChange := workerWantScaleChange || cpWantScaleChange
+
+	// The restriction that not allowing scaling and rolling is covered in webhook.
+	op := NoChange
+	switch {
+	case wantScaleChange:
+		op = ScaleOperation
+	case wantVersionChange:
+		op = K8sVersionUpgradeOperation
+	}
+	log.V(5).Info("Operation detected", "operation", op)
+	return op, nil
+}
+
+func (r *Reconciler) workerReplicasDiff(ctx context.Context, tinkerbellScope *Scope) bool {
+	workerWantScaleChange := false
+	for _, wnc := range tinkerbellScope.ClusterSpec.Cluster.Spec.WorkerNodeGroupConfigurations {
+		md := &clusterv1.MachineDeployment{}
+		mdName := clusterapi.MachineDeploymentName(tinkerbellScope.ClusterSpec.Cluster, wnc)
+		key := types.NamespacedName{Namespace: constants.EksaSystemNamespace, Name: mdName}
+		err := r.client.Get(ctx, key, md)
+		if err != nil {
```

Comment:
Does this error need to bubble up somewhere if not an `IsNotFound(err)`?

Reply:
nice catch